### DISCUSSION TITLE
Aktualitätserholung mit Schwierigkeitsgrad anpassen

### DISCUSSION THREAD

**Discussion by lonk0123** [TVTower/TVTower#789] [Ideas]
Wäre es sinnvoll die Erholungsgeschwindigkeit der Sendungen je nach gewählter Schwierigkeitsstufe anzupassen.
Mir kommt es zu einfach vor, wenn Sendungen nach 3, spätestens 4 Tagen, wieder die volle Aktualität besitzen. 
Ich würde mir einen höheren Schwierigkeitsgrad wünschen. Von mir aus kann eine Sendung auch 2 Tage mehr benötigen, bis sie wieder fit ist.

**Comment by nittka** [comment]
Ja, das klingt machbar und sinnvoll. Der Parameter der Schwierigkeitsstufe muss dann mit einem möglicherweise vorhandenen Paramter in der Lizenz kombiniert werden. Denn die kann sich auch schon von sich aus schneller oder langsamer erholen.

**Reply by nittka** [reply]
Ganz trivial scheint diese Anpassung nicht zu werden. Die Aktualität wird in den Programmdaten verwaltet, wobei sich theoretisch mehrere Lizenzen auf dieselben Programmdaten beziehen können (z.B. Einzellizenz und Film-Bundle mit beschränkter Ausstrahlungszahl im Besitz von verschiedenen Spielern). D.h. nach aktuellem Implementierungsstand gehören die Daten keinem bestimmten Spieler, so dass eine sofortige Zuordnung der Schwierigkeitsstufe nicht möglich ist.

**Comment by lonk0123** [comment]
Ich habe mir heute nochmals nach einem sehr enttäuschenden Spielstart die Aktualität angesehen.
Die Enttäuschung bestand darin, daß praktisch sämtliche Filme, die am Starttag gesendet wurden, am nächsten Tag schon wieder volle Aktualität besaßen.

Ich habe mir daher die RefreshTopicality aus game.programme.programmedata.bmx angeschaut. Dabei fiel mir auf, daß
1) diese Funktion scheinbar für alle Filme aufgerufen wird, also auch für jene, die nicht vom Spieler oder der KI gesendet worden sind, was bei zunehmender Datenbankgröße unnötige Berechnungen kostet. Da diese nicht so umfangreich sind, sicher kein Problem, aber ich wollte es erwähnen.
2) der Minimumzuwachs an Aktualität liegt momentan bei 0.1
Dies führt dazu, daß Filme mit niedriger MaxTopicality praktisch immer gleich am nächsten Tag wieder volle Aktualität besitzen.
3) der refreshModifier extrem hoch sein kann (z.B. 1.75) und dadurch Aktualitätszuwächse von 0.3 keine Seltenheit sind. Dies bedeutet, daß die meisten Filme schon am übernächsten Tag wieder volle Aktualität haben.

Ich habe daher die Funktion etwas abgeändert, so daß sich die Aktualität nun nur noch im Verhältnis zur max. Aktualität erholt, wobei pro Tag im Schnitt nicht mehr als 0.05 regeneriert werden sollen. Für Filme mit niedriger maximaler Aktualität (0.1), soll diese Regeneration entsprechend auch nur so um 0.01 liegen.

Die Funktion sieht nun (unoptimiert) folgendermaßen aus:
```
Method RefreshTopicality:Float(refreshModifier:Float = 1.0) {_private}
		Local minimumRelativeRefresh:Float = 1.1 '110%
		Local minimumAbsoluteRefresh:Float = 0.1 '10%

		refreshModifier :* GetProgrammeDataCollection().refreshFactor
		Local modifer:Float = GetRefreshModifier()
		
		If modifer <> 1.0
			refreshModifier = 1.0 + (refreshModifier - 1.0) * modifer
		EndIf
		refreshModifier :* GetGenreRefreshModifier()
		refreshModifier :* GetFlagsRefreshModifier()
		
		refreshModifier = Max(refreshModifier, minimumRelativeRefresh)
		topicality = GetTopicality() 'limit to max topicality
		
		Local temp:Float = GetMaxTopicality()
		if temp < 0.1 then
			minimumAbsoluteRefresh = 0.01
		else
			minimumAbsoluteRefresh = temp / 10
		endif
		
		topicality :+ Max(topicality * 0.1, minimumAbsoluteRefresh)
		topicality = MathHelper.Clamp(topicality, 0, GetMaxTopicality())
		Print "GetMaxTopicality = " + GetMaxTopicality()
		Return topicality
	End Method
```

Dies berücksichtigt den refreshModifier überhaupt nicht und stellt ein sehr vereinfachtes Modell dar.

Nach ersten Tests, erholt sich (mit diesem extrem vereinfachten Modell) die Aktualität wesentlich langsamer und Filme sind oftmals erst nach einer Woche (also inGame mehr als einem halben Jahr) wieder voll sendefähig, was auch der Realität mehr entsprechen sollte.

Ich werde mal einen neuen Spielstand beginnen und checken, ob man hiermit etwas anfangen kann...

**Reply by GWRon** [reply]
Danke fuers "Experimentieren".

> der Minimumzuwachs an Aktualität liegt momentan bei 0.1
> Dies führt dazu, daß Filme mit niedriger MaxTopicality praktisch immer gleich am nächsten Tag wieder volle Aktualität besitzen.

Der Gedanke dahinter ist, dass der "relative Zuwachs" bei "alten Schinken" hoeher ausfallen muss.
Wenn ich einen 80er Jahre Film _heute_ aussende, dann ist der ja schon niedrig von der (maximalen) "Aktualitaet" - und er lockt schon weniger Zuschauer vor die Kiste ("kenn ich schon" versus "guck ich immer wieder gern").
Wenn ich den Film also 10x im Jahr sende, habe ich im Ausgleich zwar weniger "gucke ich immer wieder gern", dafuer aber mehr "hach, schau ich mir den halt an, kommt ja sonst nix besseres". Die "kenn ich schon"-Gruppe war von Anfang an ja schon gross - und von ihr kommen nun mehr von denen "hach schau ich mir den halt an" dazu.

Hoffe ich konnte den Gedanken halbwegs nachvollziehbar erklaeren.


Dennoch hast Du vielleicht recht damit, dass der Erholungsfaktor zu stark ausfaellt. Hier bietet sich immer eine Mischung aus konstanter Erholung und dynamischer Erholung an.
Also sowas wie (nur symbolischer Code)
```BlitzMax
refresh = fixRefresh + dynamicRefresh '0.05 + 0.10 * (1.0 - GetMaxTopicality())
...
finalTopicality = (1 + refresh) * topicality
```
wir haetten also mindestens einen refresh von 5% und bei einem brandaktuellem Film (100% aktualitaet) bliebe es dabei. je aelter und abgenudelter der Film ist, desto groesser faellt die Auffrischung aus. In dem Beispiel jetzt reicht die Auffrischung von 5 bis 15%.

Wenn man nun noch (1.0 - GetMaxTopicality() potenziert (Exponent > 1) verstaerkt man die Anstiegskurve (muss aber dann sicher die 0.10 noch hochschrauben). Exponenten unter 1 wuerden entsprechend die Kurve abschwaechen (da das effektiv einfach nur die Wurzel zieht).

**Comment by lonk0123** [comment]
Ja, die Logik, die dahinter stehen soll, ist nachvollziehbar. 
Sie führt aber dazu, daß die alten Schinken, die halt ihre mickrige Einschaltquote erzielen, problemlos als Lückenfüller für die ersten Tage herhalten können, da immer volle Aktualität vorhanden ist.
Es reicht völlig, wenn ein alter Film die volle Aktualität erst wieder nach 2 oder 3 Tagen erreicht, und nicht schon am nächsten Tag

Im Testspielstand ergibt sich mit der neuen Refreshrate jetzt folgendes Bild nach 3 Tagen:
Man nutzt die Ramschkiste jetzt länger (sonst meist nur am Startspieltag, evtl. noch am 2. Spieltag).
Es ist schwieriger einen Kickstart hinzulegen, da die simple Wiederholungstaktik, bei der man nur Programminhalt für 2 Spieltage brauchte, nicht mehr so ohne Weiteres funktioniert.

Auf jeden Fall kann man, wenn hier eine geeignete Formel gefunden werden konnte, auch den Schwierigkeitsgrad mit unterbringen, in dem Sinne, daß es dann sowohl für den Spieler, als auch für die KI herausfordernder wird.

**Reply by GWRon** [reply]
Das Problem ist ja nur, wenn die Erholungsrate absolut ist (derzeitigeAktualitaet = derzeitigeAktualitaet + X). Wenn sie "relativ" ist, macht es am Ende natuerlich einen Unterschied ob ein Kinofilm sich um 15% seiner maximalenAktualitaet erhoeht oder um 15% seiner derzeitigenAktualitaet.

Am Ende der aktuellen `RefreshTopicality()` steht ja bereits:
```BlitzMax
		topicality :+ Max(topicality * (refreshModifier-1.0), minimumAbsoluteRefresh)
		topicality = MathHelper.Clamp(topicality, 0, GetMaxTopicality())

		Return topicality
	End Method
```

Du siehst: hier wird bereits ein Multiplikator (sprich "relative Aenderung") auf die Aktualitaet angewendet. 
Das wichtige ist, hier wird immer 0.1 als Minimum fuer die absolute Aenderung angesehen.

Das war mal eingefuehrt worden um das Problem kleiner Zahlen zu verringern: Aktualitaet von "5%" - und nun erhoehe das mal um 10%, dann landest Du bei 5,5%, davon 10% am Tag2 und du landest bei 6,05, danach knapp 6,66 ... usw. es zieht sich wie Kaugummi.

Deswegen mein Vorschlag, den Multiplikator zu potenzieren - sprich niedrige Werte brauchen staerkere Anstiege.


Edit: Ich schlage vor, dass man sich in Excel/Calc/... eine kleine Tabelle macht und dort mal mit den Werten spielt. Dann kann man sich prima Kurven anzeigen und mit Eingangswerten spielen (aktueller Film vs neuer Film usw.)

**Reply by GWRon** [reply]
![image](https://user-images.githubusercontent.com/2625226/216094460-3c24ba5f-d195-446f-9f03-f1e004b4c98e.png)

Weiter hatte ich noch keine Zeit, aber als naechster Schritt kommt die "Auffrischung" an - damit man sieht, wie lange was braucht, um sich zu erholen.

Zieht sich halt, die ganzen Funktionen in Calc "nachzubauen".

**Reply by GWRon** [reply]
Hier mal eine LibreOffice/OpenOffice-Calc-Datei

[TVTower_sim.ods](https://github.com/TVTower/TVTower/files/10569177/TVTower_sim.ods)
![image](https://user-images.githubusercontent.com/2625226/216331194-19755760-469a-4279-a4b5-51b51dbcf1d6.png)


Da sollte sich einfach mit den Werten spielen lassen. Wenn Stellschrauben am Ende auf "1,0" stehen, koennte man sie weglassen.

**Reply by GWRon** [reply]
![Peek 2023-02-02 13-59](https://user-images.githubusercontent.com/2625226/216331600-2d3040d6-c3bd-424b-90ed-cc6a1227cdd8.gif)

**Comment by lonk0123** [comment]
So, Tag 7 und man hat ein ganz neues Spielgefühl, da wesentlich höhere Anstrengungen notwendig sind, um ein einigermaßen vernünftiges Programm auf die Beine zu stellen.

Vielleicht ist man das Thema Aktualitätsrefresh auch zu komplex angegangen, da mir meine auf die Schnelle gebastelte Lösung wesentlich realistischer und auch fordernder vorkommt.

Fakt ist (in 3 Spielständen getestet), daß ein Film wie Mozarella, der am ersten Tag gesendet wird, nicht auch am 2.Tag direkt mit voller Aktualität ausgestrahlt werden darf. Solche Schummeleien machen es doch arg zu einfach.

1) Der refreshModifier fällt viel zu hoch aus, weshalb
```
topicality :+ Max(topicality * (refreshModifier-1.0), minimumAbsoluteRefresh)
```
im ersten Teil der Max-Berechnung oft zu hoch liegt. So könnte z.B. auch dieser RefreshModifier je nach Schwierigkeitsstufe unterschiedlich ausfallen bzw. nach allen Berechnungen mit einem Faktor verrechnet werden, der die Schwierigkeit repräsentiert.

2) minimumAbsoluteRefresh müßte für Produktionen niederer MaxTopicality kleiner ausfallen als 0.1, da diese sonst immer sofort am nächsten Tag wieder komplett aktuell sind (1 oder 2 Tage kann es m.M. nach ruhig dauern)

Letztlich soll ja nur dazu angeregt werden, ein differenzierteres Programm zu gestalten, statt immer nur in einer Dauerschleife von 48h die ewig gleichen Programme laufen zu lassen.

**Reply by GWRon** [reply]
Der derzeit groesste Einflussfaktor auf die "Auffrischung" ist
```

Type TProgrammeDataCollection Extends TGameObjectCollection
...
	'factor by what a programmes topicality INCREASES by a day switch
	'ex.: 1.0 = 0%, 1.5 = add 50%y
	Field refreshFactor:Float = 1.4
```
Also Basis ist hier, 40% der "aktuellen Aktualitaet" pro Tag hinzuzugewinnen - heisst, der Zuwachs sollte jeden Tag _groesser_ als am Vortag ausfallen.

Nun frage ich mich, ob wir das nicht "andersherum" machen sollten:

derzeit:
```
		topicality :+ Max(topicality * (refreshModifier-1.0), minimumAbsoluteRefresh)
```
-> mit steigender "topicality" ist der Zuwachs auch immer hoeher

Besser waere aber doch:
```
		topicality :+ Max((GetMaxTopicality()-topicality) * (refreshModifier-1.0), minimumAbsoluteRefresh)
```
Dann ist der Zuwachs jeden Tag ein wenig kleiner (bis das "minimumAbsoluteRefresh" die untere Grenze bildet)

**Reply by lonk0123** [reply]
> Besser waere aber doch:
> 
> ```
> 		topicality :+ Max((GetMaxTopicality()-topicality) * (refreshModifier-1.0), minimumAbsoluteRefresh)
> ```
> 
> Dann ist der Zuwachs jeden Tag ein wenig kleiner (bis das "minimumAbsoluteRefresh" die untere Grenze bildet)

Dies entspricht aber nicht der Realität, da das Interesse mit der Zeit wieder wachsen sollte, je länger der Film nicht ausgestrahlt wurde.
Vermutlich wären dann auch die ersten Auffrischungen des topicality-Wertes zu hoch, da die Verluste bei Ausstrahlung momentan für einige Bereiche ziemlich niedrig angesetzt sind und er wäre dann zu schnell wieder bei 80%...

**Reply by GWRon** [reply]
Doch, es sollte schon richtig sein erst "viel" dann langsamer anzusteigen.

Bedenke, dass rein theoretisch jeder Spieltag einen Monat darstellt.

Nach der Ausstrahlung ist alles erstmal im Kurzzeitgedaechtnis, man erinnert sich noch an Dialogzeilen usw.
Doch das verblasst immer weiter und weiter. 

In das Langzeitgedaechtnis wandern ikonische Momente (Terminator landet nackig / schmilzt durch Gitterstaebe, usw.).
Das Langzeitgedaechtnis merkt sich auch die generelle Geschichte ... Also die Grundpraemisse und das Ergebnis.


Diese Sache mit dem Langzeitgedaechtnis wird mit der generellen Abnutzung (maxTopicality) auch abgebildet..die sinkt ja durch jede Ausstrahlung (ab x Ausstrahlungen kein eigener Einfluss mehr auf maxTopicality).


Was man ueberlegen kann, ist die Auffrischung nicht mit einer Kurve wie derzeit stattfinden zu lassen (immer schneller bzw immer langsamer) sondern zweiteilen...erst schnell bis zu einem Plateau, dann langsam zum Ziel. Oder einen anderen Verlauf.

So kann ist das Programm nicht sofort wieder voll einsatzbereit aber als Notprogramm eher nutzbar.


> Dies entspricht aber nicht der Realität, da das Interesse mit der Zeit wieder wachsen sollte, je länger der Film nicht ausgestrahlt wurde.

Das stimmt sicher fuer viele Faelle ... das Interesse steigt und steigt. Doch irgendwann erreicht man einen Maximum (koennte die "maximale Aktualitaet" sein). Aber beinhaltet das Maximum auch folgende Situation: Jemandem gefaellt ein Film ... und nach dem Schauen gibt es ein wachsendes Gefuehl von "wuerde ich mir durchaus ein zweites Mal anschauen" - aber genau dieses Gefuehl schrumpft ja auch wieder, irgendwann ist einem der Film unwichtiger und unwichtiger und...
Frage hier: ist das bei jedem so - also kann man das auf alle Einwohner anwenden?
Eine gewisse Komplexitaet in den Formeln ist schoen und gut - aber Einfachheit hat auch Vorteile :)

**Reply by lonk0123** [reply]
Die Komplexität habe aber nicht ich eingeführt , sondern sie ist im Programmcode bisher so integriert:
Wer will schon sagen, daß Komödie einen GenreRefreshModifier von 1.25 haben muß, Western hingegen 0.95.

Genau wegen der Simplizität funktioniert mein Modell bisher auch mehr als zufriedenstellend für mich.
Es gibt einen prozentualen Anteil zurück, ohne Sonderfälle zu berücksichtigen.

Man kann halt immer alles begründen, aber mit dem erst "viel" wachsen, habe ich meine Probleme, genauso wie Du Probleme im umgekehrten Fall siehst, die ich nicht unbedingt als problematisch einstufe.
Möglicherweise kann einfach ein Wert ausgewürfelt werden, der dann zur Aktualitätssteigerung je nachdem etwas hinzuaddiert oder abzieht.

Der Spieler sieht eh nur den Balken jeden Tag ein Stück ansteigen. Die dahinterliegende Logik bleibt den meisten verborgen.

Wenn Du darauf bestehst, daß die Aktualisierung am Anfang größer sein soll, als am Ende, müßte zusätzlich auch an der CutTopicality()
geschraubt werden, da meist nicht genug Aktualität verloren geht, um dann die stärkeren Anfangszuwächse der Erholung dieser wieder auszugleichen. Die Filme wären halt zu schnell wieder im sendefähigen Bereich.

**Reply by GWRon** [reply]
Die Komplexitaet ist noetig, da sich nun mal bestimmte Genre anders "abnutzen".
Horrorfilm: die Schockmomente sind bekannt ...
Komoedie: Die Gags sind bekannt - aber in Gesellschaft lacht sichs trotzdem gut
Dramen: geheult werden kann immer
Erotik: ... die Geschichte und Dialoge sind wichtig!
...
Entsprechend gibt es einfache Modifikatoren fuer jedes Genre (damit nicht jeder Film selbst Werte definieren muss - aber halt kann)
Gleiches gilt fuer Flags (BMovie, Kultur, ...) 

Inwieweit diese Werte zu hoch, zu niedrig, ... ausfallen, muss das Balancing dann zeigen. Aus diesem Grund sind Excelsheets nicht die duemmste Idee - man kann sich in Charts/Diagrammen/... zeigen lassen, wann ein Film wieder "Einsatzbereit" waere.


> CutTopicality()

Hier ist die Sache zweigleisig zu sehen: Je weniger Leute zuschauen, desto weniger geht die Aktualitaet runter (nicht direkt proportional ...!). Heisst, die Nachtwiederholung kostet zwar ein wenig Max.Aktualitaet, aber weniger "derzeitige Aktualitaet".
Unabhaengig davon kann natuerlich die Abnutzung (WearOff im Code) auch zu hoch oder niedrig ausfallen (Balancing) und auch hier kann sich jedes Genre anders intensiv abnutzen.


Ok, warum nun zwei verschiedene Modifikatoren (Refresh und WearOff), kann nicht das eine einfach das Gegenteil vom anderen sein?
Eine Ausstrahlung "heute" senkt die Aktualitaet, eine weitere Ausstrahlung "heute", senkt sie weiter.
Die Auffrischung findet "einmal am Tag" statt. Das aber nur als Nebeninfo.
Wichtiger ist: man kann nun steuern, dass eine Ausstrahlung einen hohen Abnutzungsfaktor hat - also sich nicht gleich nach ein paar Tagen zur Wiederholung eignet, oder aber, dass sie sich zwar schnell abnutzt, aber eben auch zuegig erholt (sich also schon nach 2-3 Tagen wieder lohnt).

Ein "Kultfilm" sollte sich also schneller erholen ("wird immer wieder gern gesehen - mit zeitlichem Abstand") als ein Krimi, bei dem ja dann der Moerder noch mehr als gut im Gedaechtnis ist.

**Comment by nittka** [comment]
@lonk0123: Schön, dass Du Dich in den Code einarbeitest. (Könntest Du als nächsten Schritt einen Fork des Repo anlegen und Deine vorgeschlagenen Änderungen in einem Branch zur Verfügung stellen? Natürlich geht auch das Zusammenkopieren von Codeschipseln aus Diskussionen und Tickets, aber ein konsistenteres Bild gibt es mit einer Folge von Commits.)

Was ich mit "nicht ganz trivial" meinte war ausschließlich das Einbeziehen des Schwierigkeitslevels, da die Programmdaten derzeit keine eindeutige Zuordnung zu einem Spieler haben müssen.

Mit welchem Startjahr spielst Du? Ich vermute mal deutlich später als 1985. Könntest Du schauen, ob Deine Kaufstrategie in der unveränderten Form auch 1985 funktioniert (Kaufstrategie und Aktualitätsanpassung). Und dann mit der geänderten Formel.

**Reply by lonk0123** [reply]
> Mit welchem Startjahr spielst Du? Ich vermute mal deutlich später als 1985. Könntest Du schauen, ob Deine Kaufstrategie in der unveränderten Form auch 1985 funktioniert (Kaufstrategie und Aktualitätsanpassung). Und dann mit der geänderten Formel.

Nein, ich spiele mit 1985. 
Da ca. 75% aller eingesetzten Filme des Starttages am nächsten Tag sofort wieder volle Aktualität hatten, "mußte" ich mir etwas Neues zurechtbiegen.

@GWRon: Spar Dir erst mal die Mühe mit dem Nachbilden der Funktionen. Ich verstehe den Code mittlerweile ganz gut und hatte mir für alle Filme Beispielwerte aus den Funktionen rausgepreßt.
Ich würde noch etwas weitertesten, dann den vorhandenen Code anpassen und mit neuen Werten füttern.
Es gibt ja Funktionen wie
```
GetGenreRefreshModifier()
GetFlagsRefreshModifier()
```
die den RefreshModifier auch ganz schön pushen, die würde ich halt drin lassen wollen (Ihr habt Euch ja was dabei gedacht)
und dann schauen, daß die ganzen Steigerungen etwas milder ausfallen.

...Und ja, der
```
Field refreshFactor:Float = 1.4
```
kann dann auch je nach Schwierigkeitsgrad unterschiedliche Werte erhalten.

**Reply by nittka** [reply]
OK. dann musst Du mir mal erklären, wie Du es schaffst, ab dem zweiten/dritten Tag keine Ramschfilme mehr zu kaufen. Mein Eindruck war, dass gute Filme bei der Ausstrahlung so viel an Wert verlieren, dass die Werbeeinnahmen nicht mal den Wertverlust decken. Aber das ist ein anderes Thema.

Nochmal. Unterschiedliche Werte je Schwierigkeitsgrad sind zur Zeit eben nicht so trivial umzusetzen. Die einzelne Lizenz gehört zwar einem Spieler, aber die Programmdaten, für die die Erholung/maximale Aktualität berechnet werden, sind global für alle. Man weiß also gar nicht welcher Schwierigkeitsstufe anzusetzen ist. Man kann auch nicht die es menschlichen Spielers nehmen, da die KI ja gerade mit einer anderen Schwierigkeit spielen könnte.
Es ist prinzipiell möglich, dass zwei Spieler (mit unterschiedlichem Schwierigkeitsgrad) eine Lizenz besitzen, hinter der dieselben Programmdaten liegen.

Ein möglicher Ansatz wäre vielleicht, dass bei einer Ausstrahlung der Owner an den Programmdaten gesetzt wird (und beim Verkauf entfernt wird). Aber für die Auswirkungen stecke ich nicht tief genug in der Datenmodellierung.

**Reply by GWRon** [reply]
Wie bereits beschrieben gibt es fuer jeden "Programm" ("TProgrammeData") mindestens eine Lizenz. Die Lizenzen sind das, was die Spieler holen. Es kann mehrere Lizenzen zu einem Programm geben ... eine mit "nur 3 Ausstrahlungen", die andere mit "nur heute" usw.
Wir noch nicht wirklich genutzt - ist aber halt prinzipiell moeglich.

Einem Zuschauer ist es egal, ob er einen Film guckt, der mittels "nur 3 Ausstrahlungen"-Lizenz ausgestrahlt wurde oder nicht. Die Abnutzung erfolgt also im "TProgrammeData"-Objekt und nicht im "TProgrammeLicence"-Objekt.

Was man hier nun machen koennte - und das ist "komplizierter" ist, jedem "TProgrammeData" noch `Field users:Int` bzw `Field usedInLicencesCount:Int` einzufuehren und beim Aendern von Abnutzung/Auffrischung diesen Wert heranzuziehen.

> 1 bedeutet mehrere Lizenzen.
Bei mehreren zugeordneten Lizenzen muss eine Menge an "unterschiedlichen Spielern" generiert werden: "[1,3]" (Spieler 1 und 3 haben es). Aus den Schwierigkeitsgraden dieser Spieler werden Durchschnitte fuer den Refresh gebildet - und dieser angewendet.
Erzeugt allerdings fuer einen Spieler einen Vorteil und fuer den anderen einen Nachteil.

**Reply by nittka** [reply]
Wenn ich es richtig gesehen habe, ist TProgrammeData ein OwnedObject. Wenn man es einfach halten möchte, könnte man einfach den letzten "Ausstrahler" zum Owner für die Ermittlung des Auffrischung machen. Die Frage ist, ob das Feld schon jetzt für etwas anderes herangezogen wird - sollte eigentlich nicht, denn an sich ergibt Owner für die Daten keinen Sinn.

**Reply by GWRon** [reply]
Nein der Owner ist noch "frei". Die nutzen einfach nur aus Bequemlichkeit das gleiche "Grundobjekt".

**Reply by lonk0123** [reply]
> OK. dann musst Du mir mal erklären, wie Du es schaffst, ab dem zweiten/dritten Tag keine Ramschfilme mehr zu kaufen. Mein Eindruck war, dass gute Filme bei der Ausstrahlung so viel an Wert verlieren, dass die Werbeeinnahmen nicht mal den Wertverlust decken. Aber das ist ein anderes Thema.

Die Strategie war simple: Kaufe mit dem Startkapital (und evtl. restlichen Zusatzkredit) Filme, die so um die 20.000 pro Stunde Wert haben. Hier bieten sich insbesondere Monumentalfilme an, da die oftmals ein sehr gutes Kosten/Nutzenverhältnis besitzen (z.B. Riesenstadt Long Cut 5 Blöcke für 60.000 etc. reicht für die ersten 2 Nächte vollkommen aus, um einige kleinere Werbeverträge abzuschließen). 
Da die allermeisten Filme am nächsten Tag wieder volle Aktualität haben, braucht man also nur Programm für 1 (max. 2 Tage).
Also 48h * 20.000 --> max. 1. Mio Kosten (mit Zusatzkredit hat man schon 750.000, also kein Ding)
Damit ist die Sendeschleife komplett und fast alle Filme können mit max. Aktualität nun im 48h-Rhythmus  immer wiederholt werden.

Selbst gute Filme, wie Mozarella (glaube 2h für 80.000) haben am nächsten Tag wieder volle Aktualität gehabt, also der Wertverlust, den Du siehst, gibt es ja meist gar nicht. Er wäre ja auch nur temporär, also 1, max. 2 Tage vorhanden.

Die Kosten sind also nach Tag 1 = 0! (Selbst der Startsender kostet keinen Unterhalt)
Die Werbeeinnahmen bleiben jetzt direkt beim Spieler hängen, da wie gesagt keinerlei Ausgaben (Nachrichtenkosten vernachlässigt)

Nun kann man ab Tag 2 verschiedene Strategien fahren, als da wären:
1) Senderausbau
2) Verbesserung der Programminhalte
3) Gemischt aus 1 + 2

Deshalb war es für mich so enorm wichtig, die Aktualitätsregeneration nach unten anzupassen, da diese Sendeschleifen-Strategie bisher immer funktioniert hat. Da sich die Filme nach meiner Formel jetzt erst nach 7-10 Tagen wieder voll regeneriert haben, ist die Sendeschleife auch vom Umfang dementsprechend angestiegen.

Noch ein Wort zum Schwierigkeitsgrad bei der Anpassung der Aktualitätserholung:
Nachdem ich gestern mit Ron gesprochen habe und er mich nochmals auf das zugrunde liegende Problem der Zuordenbarkeit der Lizenzen aufmerksam machte, möchte ich bemerken, daß es mir nur darum ging, den Schwierigkeitsgrad des Spieles allgemein zu steigern, nicht im Verhältnis zur KI.
Wir sind überein gekommen, daß dies optional in den Einstellungen geschehen könnte. 
Aber ich will Deiner Konzeptausarbeitung nicht vorgreifen, da dies nur ein Einzelschritt im Gefüge zahlreicher Stellschrauben wäre...

**Comment by nittka** [comment]
Ich bin leider mit dem Konzeptentwurf noch nicht weitergekommen. Ich will deshalb nur kurz anmerken, dass die Aktualitätserholung und die maximale Aktualität auch nur ein Teil des Gesamtkonstrukts sind. Reichweite und Image sind Faktoren, die auch nicht vernachlässigt werden dürfen.

Am Ende geht es doch darum (glaube ich zumindest)
* ein Spieler muss Filme zeigen, um Werbeeinnahmen zu generieren (Ausbau von Fundus und Reichweite, Decken der laufenden Kosten)
* man will verhindern, dass es ausreicht, immer wieder dieselben Filme zu senden (Werbeeinnahmen für 0 Aufwand)

Wenn die Einschaltquote noch stärker an die Ausstrahlungshäufigkeit gekoppelt wäre, könnte man schon deshalb nicht denselben Kram immer wieder zeigen. Diese Kopplung ist später im Spiel aber wichtiger als am Anfang. Je höher das Image ist, desto höheren Anspruch haben die Leute an das Programm. Je größer die mögliche Zuschauerzahl ist (desto mehr Werbeeinnahmen stehen zur Verfügung und desto größer dürfte der Fundus ohnehin schon sein), desto länger braucht es, bis sich ein Film erholt.
Wenn die Imageänderung nicht allein an der Einschaltquote festgemacht wird, sondern Sendezeit und Ausstrahlungshäufigkeit berücksichtigt (es ist nicht so schlimm, Wiederholungen im Morgenprogramm zu senden, aber die vierte Wiederholung zur Prime-Time bedeutet Punktabzug), steigt ggf. auch das Image nicht so schnell und das Ausbau für Kabel und Satellit dauert länger...

Die Einzelparameter zu optimieren ist ein wichtiger Punkt. Aber ich würde gern auch das Gesamtbild zusammengefasst haben. Eine Liste der Punkte, die aktuell nicht optimal sind (was stört uns), eine Liste von Zielen die erreicht werden soll (welche Faustregeln sollten gelten), eine Liste von "globalen" Maßnahmen, die man ergreifen kann.
Ich denke erst mit diesem Gesamtbild kann man dann prüfen, ob die Änderungen das Gewünschte erreicht haben.
